drivers/led/lp5562.c

```diff
@@ -0,0 +1,879 @@
+ */
+static inline int lp5562_program_wait(struct device *dev,
+				      enum lp5562_led_sources engine,
+				      u8_t command_index,
+				      u32_t time)
+{
+	/*
+	 * A wait command is a ramp with the step_count set to 0. The fading
+	 * direction does not matter in this case.
+	 */
+	return lp5562_program_ramp(dev, engine, command_index,
+			time, 0, LP5562_FADE_UP);
+}
+
+/*
+ * @brief Program a command to go back to the beginning of the program.
+ *
+ * Can be used at the end of a program to loop it infinitely.
+ *
+ * @param dev           LP5562 device.
+ * @param engine        Engine to be programmed.
+ * @param command_index Index of the command in the program sequence.
+ *
+ * @retval 0       On success.
+ * @retval -EINVAL If the given command index is out of range or an invalid
+ *		   engine is passed.
+ * @retval -EIO    If the underlying I2C call fails.
+ */
+static inline int lp5562_program_go_to_start(struct device *dev,
+					     enum lp5562_led_sources engine,
+					     u8_t command_index)
+{
+	return lp5562_program_command(dev, engine, command_index, 0x00, 0x00);
+}
+
+static int lp5562_led_blink(struct device *dev, u32_t led,
+			    u32_t delay_on, u32_t delay_off)
+{
+	struct lp5562_data *data = dev->driver_data;
+	struct led_data *dev_data = &data->dev_data;
+	int ret;
+	enum lp5562_led_sources engine;
+
+	ret = lp5562_get_available_engine(dev, &engine);
+	if (ret) {
+		return ret;
+	}
+
+	ret = lp5562_set_led_source(dev, led, engine);
+	if (ret) {
+		SYS_LOG_ERR("Failed to set LED source.");
+		return ret;
+	}
+
+	ret = lp5562_set_engine_op_mode(dev, engine, LP5562_OP_MODE_LOAD);
+	if (ret) {
+		return ret;
+	}
+
+	ret = lp5562_program_set_brightness(dev, engine, 0,
+			dev_data->max_brightness);
+	if (ret) {
+		return ret;
+	}
+
+	ret = lp5562_program_wait(dev, engine, 1, delay_on);
+	if (ret) {
+		return ret;
+	}
+
+	ret = lp5562_program_set_brightness(dev, engine, 2,
+			dev_data->min_brightness);
+	if (ret) {
+		return ret;
+	}
+
+	ret = lp5562_program_wait(dev, engine, 3, delay_off);
+	if (ret) {
+		return ret;
+	}
+
+	ret = lp5562_program_go_to_start(dev, engine, 4);
+	if (ret) {
+		return ret;
+	}
+
+	ret = lp5562_start_program_exec(dev, engine);
+	if (ret) {
+		SYS_LOG_ERR("Failed to execute program.");
+		return ret;
+	}
+
+	return 0;
+}
+
+static int lp5562_led_set_brightness(struct device *dev, u32_t led, u8_t value)
+{
+	struct lp5562_data *data = dev->driver_data;
+	struct led_data *dev_data = &data->dev_data;
+	int ret;
+	u8_t val, reg;
+	enum lp5562_led_sources current_source;
+
+	if ((value < dev_data->min_brightness) ||
+			(value > dev_data->max_brightness)) {
+		return -EINVAL;
+	}
+
+	ret = lp5562_get_led_source(dev, led, &current_source);
+	if (ret) {
+		return ret;
+	}
+
+	if (current_source != LP5562_SOURCE_PWM) {
+		if (lp5562_is_engine_executing(dev, current_source)) {
+			/*
+			 * API expectes set_brightness to have no effect if the
+			 * LED is blinking already.
+			 */
```

Comment:
This comment is not true! API states that,`set_brightness` should have no effect on the blinking rate if called after `led_blink`.

Reply:
ouh, than I misinterpreted that:
``Calling this function after led_blink() won't affect blinking.``
I thought this means the blinking is not changed at all. So what should happen is that the LED keeps blinking with the previously defined rate but the brightness should change?

Reply:
yes, only the brightness is expected to change but the LED should continue blinking at the same rate.